packages/service-library/src/servicelib/redis/_decorators.py

```diff
@@ -1,56 +1,148 @@
 
-            redis_client = redis(*args, **kwargs) if callable(redis) else redis
-            assert isinstance(redis_client, RedisClientSDK)  # nosec
+            client = (
+                redis_client(*args, **kwargs)
+                if callable(redis_client)
+                else redis_client
+            )
+            assert isinstance(client, RedisClientSDK)  # nosec
+            nonlocal lock_value
+            if lock_value is None:
+                lock_value = f"locked since {arrow.utcnow().format()} by {client.client_name} on {socket.gethostname()}"
 
-            async with redis_client.lock_context(
-                lock_key=redis_lock_key, lock_value=lock_value
+            lock = client.create_lock(redis_lock_key, ttl=DEFAULT_LOCK_TTL)
+            if not await lock.acquire(
+                token=lock_value,
+                blocking=blocking,
+                blocking_timeout=(
+                    blocking_timeout.total_seconds() if blocking_timeout else None
+                ),
             ):
-                return await func(*args, **kwargs)
+                raise CouldNotAcquireLockError(lock=lock)
+
+            try:
+                async with asyncio.TaskGroup() as tg:
+                    started_event = asyncio.Event()
+                    # first create a task that will auto-extend the lock
+                    auto_extend_lock_task = tg.create_task(
+                        _periodic_auto_extender(lock, started_event),
+                        name=_EXCLUSIVE_AUTO_EXTEND_TASK_NAME.format(
+                            redis_lock_key=redis_lock_key
+                        ),
+                    )
+                    # NOTE: In case the work thread is raising right away,
+                    # this ensures the extend task ran once and ensure cancellation works
+                    await started_event.wait()
+
+                    # then the task that runs the user code
+                    assert asyncio.iscoroutinefunction(func)  # nosec
+                    work_task = tg.create_task(
+                        func(*args, **kwargs),
+                        name=_EXCLUSIVE_TASK_NAME.format(func_name=func.__name__),
+                    )
+
+                    res = await work_task
+                    auto_extend_lock_task.cancel()
+                    return res
+
+            except BaseExceptionGroup as eg:
+                # Separate exceptions into LockLostError and others
+                lock_lost_errors, other_errors = eg.split(LockLostError)
+
+                # If there are any other errors, re-raise them
+                if other_errors:
+                    assert len(other_errors.exceptions) == 1  # nosec
+                    raise other_errors.exceptions[0] from eg
+
+                assert lock_lost_errors is not None  # nosec
+                assert len(lock_lost_errors.exceptions) == 1  # nosec
+                _logger.error(  # noqa: TRY400
```

Comment:
THOUGHT: I am not sure `_logger.error` will capture in this case the traceback provided that it comes from another thread and captured in `lock_lost_errors.exceptions`. Perhaps you can pass it in the message?

Reply:
but the traceback is activated with `error`, right? ... wait, perhaps it was `exception`. All good! :-)